File: packages/core/src/parser/__testdata__/sample.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const max_retries: u32 = 3;
pub const default_model = "claude-opus-4-6";
const buffer_size: usize = 4096;

pub const Status = enum {
    active,
    idle,
    stopped,
    err,
};

pub const AgentConfig = struct {
    model: []const u8 = default_model,
    max_tokens: u32 = 4096,
    temperature: f64 = 0.7,
};

pub const AgentError = error{
    InvalidMessage,
    ModelError,
    Timeout,
};

pub const Agent = struct {
    name: []const u8,
    config: AgentConfig,
    status: Status,
    allocator: Allocator,

    const Self = @This();

    pub fn init(allocator: Allocator, name: []const u8, config: ?AgentConfig) Self {
        return .{
            .name = name,
            .config = config orelse AgentConfig{},
            .status = .idle,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *Self) void {
        _ = self;
    }

    pub fn process(self: *Self, message: []const u8) AgentError![]const u8 {
        if (message.len == 0) return AgentError.InvalidMessage;
        self.status = .active;
        defer self.status = .idle;
        return self.callModel(message);
    }

    pub fn getTools(self: Self) [3][]const u8 {
        _ = self;
        return .{ "search", "read", "write" };
    }

    fn callModel(self: Self, message: []const u8) []const u8 {
        _ = self;
        // TODO: implement actual model call
        return message;
    }

    fn validate(self: Self, input: []const u8) bool {
        _ = self;
        return input.len > 0;
    }
};

pub fn createAgent(allocator: Allocator, name: []const u8) Agent {
    return Agent.init(allocator, name, null);
}

test "agent creation" {
    const agent = createAgent(std.testing.allocator, "test-agent");
    try std.testing.expectEqual(Status.idle, agent.status);
}

// FIXME: memory management for response strings

File: packages/core/src/parser/__testdata__/sample.swift
import Foundation
import Combine

let maxRetries = 3
let defaultModel = "claude-opus-4-6"

enum Status: String, Codable {
    case active, idle, stopped, error
}

struct AgentConfig: Codable {
    var model: String = defaultModel
    var maxTokens: Int = 4096
    var temperature: Double = 0.7
}

protocol Agent: AnyObject {
    var name: String { get }
    var status: Status { get }
    func process(_ message: String) async throws -> String
    func getTools() -> [String]
}

protocol Configurable {
    func configure(_ config: AgentConfig)
}

class BaseAgent: Agent {
    let name: String
    private(set) var status: Status = .idle
    var config: AgentConfig

    init(name: String, config: AgentConfig = AgentConfig()) {
        self.name = name
        self.config = config
    }

    func process(_ message: String) async throws -> String {
        fatalError("Must override")
    }

    func getTools() -> [String] { [] }

    func validate(_ input: String) -> Bool {
        !input.trimmingCharacters(in: .whitespaces).isEmpty
    }
}

class SynapseAgent: BaseAgent, Configurable {
    private let queue = DispatchQueue(label: "synapse.agent")

    override func process(_ message: String) async throws -> String {
        guard validate(message) else {
            throw AgentError.invalidInput("Empty message")
        }
        status = .active
        defer { status = .idle }
        return await callModel(message)
    }

    override func getTools() -> [String] {
        ["search", "read", "write"]
    }

    func configure(_ config: AgentConfig) {
        self.config = config
    }

    private func callModel(_ message: String) async -> String {
        // TODO: implement actual API call
        "Response to: \(message)"
    }

    static func create(name: String) -> SynapseAgent {
        SynapseAgent(name: name)
    }
}

enum AgentError: Error, LocalizedError {
    case invalidInput(String)
    case timeout
    case modelError(String)

    var errorDescription: String? {
        switch self {
        case .invalidInput(let msg): return "Invalid input: \(msg)"
        case .timeout: return "Request timed out"
        case .modelError(let msg): return "Model error: \(msg)"
        }
    }
}

extension SynapseAgent: CustomStringConvertible {
    var description: String { "\(name) [\(status.rawValue)]" }
}

// FIXME: race condition on status property
